Use ES aliases instead of index names in queries

sravfeyn opened the pull request:
This is only open to get any feedback and not ready for merge.

We are currently querying ES via index names. Querying via alias is supported and will be needed for ICDS where we are thinking to have multiple indices under one alias. This PR doesn't implement any multiple index support, but as a starting point simply queries ES using alias instead of index in all places that I could find. I have added a logger to find any places that I missed where an index is directly queried instead of an alias.

My plan for this PR is
- to get tests passing
- test it on staging with `ENABLE_ES_INTERFACE_LOGGING`
- if those two go okay, I will open this PR for merge.  

After this my plan to actually add multiple index support for single alias. My tentative plan is
 
- add an additional `write_alias` attribute to `ElasticsearchIndexInfo` and use that for all write operations.
- add a validation to check that `write_alias` points to only single index
- rename current `alias` attribute of `ElasticsearchIndexInfo` to read_alias and use that for all `read` operations.

This will enable individual envs to manage their ES indices outside of HQ with custom settings, rolling indices etc

sravfeyn committed:
No need for EsMeta since ucr elastic backend is gone

sravfeyn committed:
better var name

sravfeyn committed:
use alias instead of es index

sravfeyn committed:
use es alias in ESView

sravfeyn committed:
use alias in escheck

sravfeyn committed:
more places

sravfeyn committed:
update tests to use alias

sravfeyn committed:
use alias

sravfeyn committed:
update arg names

sravfeyn committed:
pass alias in bulk ops

sravfeyn committed:
unused import

sravfeyn committed:
log missed index use cases

snopoke commented on corehq/elastic.py:
```diff
@@ -154,24 +152,22 @@ def send_to_elasticsearch(index_name, doc, delete=False, es_merge_update=False):
+    "sms": SMS_INDEX_INFO,
+    "report_cases": REPORT_CASE_INDEX,
+    "report_xforms": REPORT_XFORM_INDEX,
+    "case_search": CASE_SEARCH_INDEX_INFO,
```
:heart:

snopoke commented on corehq/util/es/interface.py:
```diff
@@ -9,6 +10,15 @@ class AbstractElasticsearchInterface(metaclass=abc.ABCMeta):
+            logger = logging.getLogger('es_interface')
+            all_es_aliases = [index_info.alias for index_info in ES_META.values()]
+            if index_or_alias not in all_es_aliases:
+                logger.info("Found a usecase where index is queries instead of alias")
```
```suggestion
                logger.info("Found a use case where index is queries instead of alias: %s", index_or_alias)
```

sravfeyn replied:
(Done)

snopoke commented on corehq/util/es/interface.py:
```diff
@@ -9,6 +10,15 @@ class AbstractElasticsearchInterface(metaclass=abc.ABCMeta):
+            all_es_aliases = [index_info.alias for index_info in ES_META.values()]
+            if index_or_alias not in all_es_aliases:
+                logger.info("Found a usecase where index is queries instead of alias")
+                logger.info(traceback.print_stack())
```
```suggestion
                logger.info(traceback.format_stack())
```

snopoke commented on corehq/util/es/interface.py:
```diff
@@ -61,6 +77,7 @@ def bulk_ops(self, actions, stats_only=False, **kwargs):
         return bulk(self.es, actions, stats_only=stats_only, **kwargs)
 
     def search(self, index=None, doc_type=None, body=None, params=None, **kwargs):
+        self._verify_is_alias(index)
```
rename `index` to `index_alias` to be consistent with other methods

snopoke reviewed (commented):
Looks good @sravfeyn

sravfeyn committed:
comments

sravfeyn committed:
merge master

sravfeyn committed:
import

sravfeyn committed:
fix removed class in tests

sravfeyn committed:
fix test

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
prefix alias as well; rename method

sravfeyn committed:
add option to set alias; and let all tests set alias

sravfeyn committed:
same as other ESQuery

sravfeyn committed:
fix arg name

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
fix few things overlooked earlier

sravfeyn committed:
list

sravfeyn committed:
prefix issue

sravfeyn committed:
update test alias

sravfeyn committed:
initialize index also alias

sravfeyn committed:
full index setup

sravfeyn committed:
full init

sravfeyn committed:
rename

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
Use index in bulk ops called from reindexers; convert reindexers to bulk

sravfeyn commented:
@snopoke  @dannyroberts 

FYI, I hit a small blocker on this. It impacts the deploy/reindex workflow, so I am flagging you both

The issue is that this PR makes the reindexer code also use aliases instead of index names. Since our workflow when editing an Elasticsearch mapping is to only create a new index in a private release, but not assign any alias, the reindexer code writes to old index instead of new because it continues to use the old alias pointing to old  index. There are two options here to resolve this. 

One is to update our workflow of changing ES mapping to also assign a separate ES alias, let's say, xforms_new pointing to new index, this way the reindexer will use the new alias pointing to new index. Once index is ready, flip the alias xforms from old index to new index (which should happen automatically in the code deploy which has flip_all command).

Second option is to make all the reindexer code to explicitly use index name instead of alias. The issue with this is that it needs a flag to be passed in the call chain of Reindexer -> Pillow_Or_processor -> ElasticsearchInterface. Luckily, this needs change in [just one place](https://github.com/dimagi/commcare-hq/pull/26733/commits/1844cdb9d86b3c3c5435a167cb0543077fb4cf00#diff-37cbd762f6bb728a469145592aff76b0R211) for reindexers that use ResumableBulkElasticPillowReindexer. So I converted some reindexers that don't use that reindexer to use it in this commit https://github.com/dimagi/commcare-hq/pull/26733/commits/1844cdb9d86b3c3c5435a167cb0543077fb4cf00 but there are two more Reindexers which need a lot more effort to refactor to use ResumableBulkElasticPillowReindexer .

So, I am leaning to go with option 1 since that avoids using index-name at all and is much minimal change than the 2nd option. What do you guys think?

sravfeyn committed:
Convert report form/case reindexers too

snopoke commented:
> but there are two more Reindexers which need a lot more effort to refactor to use ResumableBulkElasticPillowReindexer.

which ones are these?

sravfeyn committed:
pass iteration_key

sravfeyn committed:
Don't auto create indices while inserting docs

sravfeyn committed:
Revert "pass iteration_key"

This reverts commit a76cabec47fc64b4324bb694d17bb42a9dfbe14e.

sravfeyn committed:
Revert "Convert report form/case reindexers too"

This reverts commit c785e32b018e8396f3309ec56256057b1a3a6bf0.

sravfeyn committed:
Let ES reindexers initialize ES alias also

sravfeyn committed:
update docs

sravfeyn commented:
> which ones are these?
The two are ReportCaseReindexerFactory and the ReportFormReindexerFactory. I have tried this in one of the latest commits, but reverted as it's quite more complicated than that.

> My initial feeling is that we should change the reindexers to use the index names directly however that could be wrong if the aliases get set up with lifecycle / rollover policies in which case you'd want to write to the alias directly. So I think you may be right that creating a new alias when reindexing is the way to go.

Yeah, I think this feels the right way, I have updated the reindexer code to set alias also and use it. I have updated the relevant ES reindexing docs to reflect this.

sravfeyn committed:
import

sravfeyn committed:
no need for skip_alias param

sravfeyn committed:
no set_alias

sravfeyn committed:
test debug

sravfeyn committed:
convert to use ES interface

sravfeyn committed:
fail early with debug data

sravfeyn committed:
args

sravfeyn committed:
simple assert

sravfeyn committed:
debug

sravfeyn commented:
@dannyroberts This will affect all envs, can you confirm if this is okay for other envs as well? Docs https://github.com/dimagi/commcare-hq/pull/26733/files#diff-55ae021bb3841b645117710c304ad227

sravfeyn committed:
fix

dannyroberts commented on docs/elasticsearch.rst:
```diff
@@ -63,9 +65,15 @@ expected to return a ``doc_dict``, so just add the ``username`` to that.
-that new mapping, so you'll have to update the hash at the top of the file.
-This can just be a random alphanumeric string.  This will trigger a preindex
-as outlined in the `Indexes` section.
+that new mapping, updating index name in the mapping file triggers HQ to
+create the new index with new mapping and reindex all data, so you'll
+have to update the index hash and alias at the top of the mapping file.
```
nit
```suggestion
that new mapping. Updating the index name in the mapping file triggers HQ to
create the new index with the new mapping and reindex all data, so you'll
have to update the index hash and alias at the top of the mapping file.
```

dannyroberts replied:
nit: bump

dannyroberts commented on docs/elasticsearch.rst:
```diff
@@ -63,9 +65,15 @@ expected to return a ``doc_dict``, so just add the ``username`` to that.
+likewise or simply suffixed with ``_new``, for example ``xforms_new``. This will
+trigger a preindex as outlined in the `Indexes` section. In subsequent commits
+alias can be flipped back to what it was, for example ``xforms``. Changing the alias
+name doesn't trigger a reindex.
```
I'm kind of just catching up here. Can you explain why we need to change the alias as well as the index name? I thought the whole point of using an alias and an index name was that the index name could change while keeping the alias the same

dannyroberts replied:
OK sorry, I see the large comment above about this.

Looks like this also felt wrong to you guys but you couldn't figure out a way to use the index that still had the right behavior?

sravfeyn replied:
Yeah, this doesn't feel completely right but is better than other options. This is discussed in [comments following this](https://github.com/dimagi/commcare-hq/pull/26733#issuecomment-610909316)

All of the current code uses index names directly rather than aliases, this PR starts making use of the aliases. But our index mapping change workflow includes building a new index with HQ itself, so without creating a separate alias during the index phase, there isn't any way to prevent writing to the old index itself.

snopoke replied:
@dannyroberts I don't think it's that we couldn't figure out a way to use the index. An alias is designed to allow you to run searches across multiple indexes which is what we'd like to start using it for (e.g. monthly form indexes). We have been using it in CommCare as a way to 'flip' from one index to another after reindex (though we haven't actually used the aliases in code for a long time).

sravfeyn committed:
debug on every client

sravfeyn committed:
404 error

sravfeyn committed:
debug in interface

sravfeyn committed:
more debug

sravfeyn committed:
+more debug

sravfeyn committed:
++more

sravfeyn committed:
+++more

sravfeyn committed:
++++more

sravfeyn committed:
+5 more debug

sravfeyn committed:
+6 debug

sravfeyn committed:
auto index falsE

sravfeyn committed:
merge sr-es-auto

sravfeyn committed:
test needs user index

sravfeyn committed:
Remove all debug stuff

sravfeyn committed:
remove other debug info

sravfeyn committed:
no new line

sravfeyn committed:
fix kw

sravfeyn committed:
init index

sravfeyn committed:
import

sravfeyn committed:
no need for index based bulk op

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn commented:
I have tested this on staging by loading few ES dependent features and making sure that only aliases are queried and not indices (by looking at HTTP logs on ES obtained by tcpdumps and making sure that this [soft-assert log](https://github.com/dimagi/commcare-hq/pull/26733/files#diff-35449daeec5614e63f3a53d79459660aR14) isn't triggered). So this is ready for merge pending any further code review. I am going to convert that soft assert log into a hard failure in a separate PR in future after a week of this being live on other envs without any issues.

All of HQ with this PR uses aliases for reading/writing, and uses direct index names for index management and pillow checkpointing.

Pinging for a final code review on this @dannyroberts @esoergel @proteusvacuum @orangejenny  (cc @snopoke)

sravfeyn commented on corehq/apps/es/case_search.py:
```diff
@@ -31,7 +31,7 @@
 
 class CaseSearchES(CaseES):
-    index = CASE_SEARCH_ALIAS
+    index = "case_search"
```
It could be. This should be one of the keys in `ES_META`, it's neither an alias, nor and index name. Since `CASE_SEARCH_ALIAS` may no longer evaluate to this value, I had to change it. It's consistent with all other `HQESQuery` classes, so I let this be.

sravfeyn commented on corehq/apps/domain/tests/test_domain_calculated_properties.py:
```diff
@@ -45,7 +46,7 @@ def _set_up_sms_es(self):
         }
         send_to_elasticsearch(
-            index=SMS_INDEX_INFO.index,
+            alias=SMS_INDEX_INFO.index,
```
It shouldn't be an issue, since the test is not about ES, but to be consistent, I will update it

sravfeyn committed:
use alias

sravfeyn commented on corehq/ex-submodules/pillowtop/es_utils.py:
```diff
@@ -137,14 +140,15 @@ def completely_initialize_pillow_index(pillow):
     This utility can be used to initialize the elastic index and mapping for a pillow
     """
-    return initialize_index_and_mapping(pillow.get_es_new(), get_index_info_from_pillow(pillow))
+    return initialize_index_and_mapping(pillow.get_es_new(), get_index_info_from_pillow(pillow), set_alias=False)
```
Yeah, missed that. Actually the parent function is an orphan, so I will remove it altogether.

sravfeyn committed:
remove unused function

sravfeyn committed:
no assume_alias needed

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
ES auto-create-index to False again

sravfeyn committed:
Dont retry in tests

sravfeyn committed:
try once

sravfeyn committed:
Merge branch 'sr-auto-es' into sr-es-alias

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn committed:
test timeout faster

sravfeyn commented:
Hey guys  @snopoke @esoergel @dannyroberts , can I get a review on this (I think you have already reviewed most of it). This was blocked due to tests taking lot of time due to the auto-create-index I added previously. I have now fixed that issue in [this other PR](https://github.com/dimagi/commcare-hq/pull/27614) so this is now good to go pending a review.

dannyroberts commented on corehq/apps/receiverwrapper/tests/__init__.py:
```diff
@@ -6,4 +6,4 @@
     'get_version_from_appversion_text': get_version_from_appversion_text,
     'get_commcare_version_from_appversion_text': get_commcare_version_from_appversion_text,
-}
+}
```
nit: doesn't seem like this needs to be in the changeset

dannyroberts commented on docs/elasticsearch.rst:
```diff
@@ -20,11 +20,13 @@ domains, and they store additional mappings allowing you to query on form
 ``xforms_1cce1f049a1b4d864c9c25dc42648a45``.  Each type of index has an alias
 with the short name, so you should normally be querying just ``xforms``, not
-the fully specified index+hash.
+the fully specified index+hash. All of HQ code except the index maintaianance
```
nit: bump

dannyroberts reviewed (approved):
Had a few small comments, but overall it looks good

dannyroberts commented:
Looks like there are still issues with the tests on this PR and that's what the other PR's supposed to solve? Seems worth merging that other PR first and the merging master into this and making sure tests pass after that?

sravfeyn commented:
Thanks Danny. I made another PR that should fix that issue https://github.com/dimagi/commcare-hq/pull/27849/.

sravfeyn committed:
pr feedback

sravfeyn commented:
Build passed. Testing one last time on staging before merging.

sravfeyn committed:
Remove unused names from ES_META

sravfeyn committed:
introduce code specific index names to replace aliases that might change

sravfeyn committed:
update pillow reindex map to use index names

sravfeyn committed:
use previouslu being used names to simplify

sravfeyn committed:
update names in tests

sravfeyn committed:
add a todo note for future

sravfeyn committed:
Merge branch 'master' into sr-es-alias

sravfeyn commented:
@snopoke @dannyroberts I have added few commits to fix an issue in rolling out an index change that require a review.

sravfeyn committed:
remove import

sravfeyn committed:
undo user es change and add test-index name

sravfeyn committed:
Remove mistaken self

sravfeyn commented:
Bumping for review on recently added commits.

sravfeyn commented:
@snopoke @dannyroberts Bumping again for review on last 3 commits

snopoke commented on corehq/ex-submodules/pillowtop/tests/utils.py:
```diff
@@ -33,7 +34,7 @@ def get_doc_count(es, index, refresh_first=True):
         es.indices.refresh(index)
     stats = es.indices.stats(index)
-    return stats['indices'][index]['total']['docs']['count']
+    return list(stats['indices'].values())[0]['total']['docs']['count']
```
does this work for both versions?

snopoke commented on docs/elasticsearch.rst:
```diff
@@ -63,9 +65,16 @@ expected to return a ``doc_dict``, so just add the ``username`` to that.
+production operations continue to use the old alias pointing to existing index.
+This will trigger a preindex as outlined in the `Indexes` section. In subsequent commits
+alias can be flipped back to what it was, for example ``xforms``. Changing the alias
+name doesn't trigger a reindex.
```
when the alias is renamed in code will that automatically get updated in ES? Will the old (`xforms_new`) alias still exist or does it get removed?

snopoke replied:
If the old alias remains then I would suggest a different suffix other than `_new` like the date or something

sravfeyn replied:
Good idea, I have updated the docs.

snopoke replied:
I don't see the changes

sravfeyn commented:
@dannyroberts @snopoke Is this good to go?

snopoke commented:
@sravfeyn still some outstanding comments

sravfeyn commented:
@snopoke Sorry, which ones Simon?

snopoke commented:
@sravfeyn https://github.com/dimagi/commcare-hq/pull/26733#discussion_r448228772 and https://github.com/dimagi/commcare-hq/pull/26733#discussion_r448221878

sravfeyn commented:
I am not sure why you are not able to see my responses made couple days back.

<img width="878" alt="Screenshot 2020-07-07 at 4 58 11 PM" src="https://user-images.githubusercontent.com/829142/86773274-17208280-c073-11ea-829d-d0ee31c1e31e.png">

snopoke commented:
@sravfeyn thanks for the responses, your comments are still pending, you need to click 'submit review' (I've been bitten by that many times).

sravfeyn committed:
Update elasticsearch.rst

sravfeyn commented:
> you need to click 'submit review' (I've been bitten by that many times).
Ah, thanks!

sravfeyn commented:
@snopoke I have responded to the comments (hope you are able to see them) and update the docs per your comment.

snopoke commented:
@sravfeyn have these changes been tested on staging at all?

sravfeyn commented:
This is in QA now

sravfeyn commented:
QA is passed on this https://dimagi-dev.atlassian.net/browse/QA-1532

@snopoke think we can merge this now?

sravfeyn commented:
Hello @dimagi/team-commcare-hq 

This is a pretty big change in Elasticsearch, so wanted to make an announcement. From this PR on, HQ code will use Elasticsearch aliases to interact with Elasticsearch instead of index names as it previously did. There is also a change in the workflow of changing ES index mappings, please read it here https://github.com/dimagi/commcare-hq/pull/26733/files#diff-55ae021bb3841b645117710c304ad227

This is QA and review pass. I am going to merge it tomorrow (to give some time for people to be aware and raise any concerns if any)

sravfeyn committed:
merge master

sravfeyn committed:
Update es.py

sravfeyn commented:
This is well tested but just to be safe, if any issues with pillows do come up form this PR, this can be reverted. We just need to rewind pillows https://github.com/dimagi/commcare-hq/pull/13154/files#diff-5e1c4d805a2ca1169c83e9e2446c318eR76
